hedera-node/src/main/java/com/hedera/services/txns/schedule/ScheduleProcessing.java

```diff
@@ -0,0 +1,288 @@
+	private final long maxProcessingLoopIterations;
+
+	SigMapScheduleClassifier classifier = new SigMapScheduleClassifier();
+	SignatoryUtils.ScheduledSigningsWitness signingsWitness = SignatoryUtils::witnessScoped;
+	Predicate<ScheduleVirtualValue> isFullySigned;
+
+	@Inject
+	public ScheduleProcessing(final SigImpactHistorian sigImpactHistorian, final ScheduleStore store,
+			final ScheduleExecutor scheduleExecutor, final GlobalDynamicProperties dynamicProperties,
+			final ScheduleSigsVerifier scheduleSigsVerifier,
+			@ScheduleThrottle final TimedFunctionalityThrottling scheduleThrottling,
+			final Supplier<MerkleScheduledTransactions> schedules) {
+		this.sigImpactHistorian = sigImpactHistorian;
+		this.store = store;
+		this.scheduleExecutor = scheduleExecutor;
+		this.dynamicProperties = dynamicProperties;
+		this.scheduleThrottling = scheduleThrottling;
+		isFullySigned = scheduleSigsVerifier::areAllKeysActive;
+		this.schedules = schedules;
+		this.maxProcessingLoopIterations = dynamicProperties.schedulingMaxTxnPerSecond() * 2;
+	}
+
+	/**
+	 * Expires all scheduled transactions, that are in a final state, having an expiry before consensusTime,
+	 * and having an expiry before any transaction that is ready to execute.
+	 */
+	public void expire(Instant consensusTime) {
+
+		// we _really_ don't want to loop forever. If the database isn't working, then it's possible. So we put
+		// an upper limit on the number of iterations here.
+		for (long i = 0; i < maxProcessingLoopIterations; ++i) {
+			store.advanceCurrentMinSecond(consensusTime);
+
+			List<ScheduleID> txnIdsToExpire = store.nextSchedulesToExpire(consensusTime);
+			if (txnIdsToExpire.isEmpty()) {
+				return;
+			}
+
+			for (var txnId : txnIdsToExpire) {
+				store.expire(txnId);
+				sigImpactHistorian.markEntityChanged(fromScheduleId(txnId).longValue());
+			}
+		}
+
+		log.error("Reached maxProcessingLoopIterations reached in expire, we should never get here!");
+	}
+
+	/**
+	 * Gets the next scheduled transaction that is available to execute. Scheduled transactions may be expired
+	 * as needed during this call.
+	 *
+	 * @param consensusTime the current consensus time
+	 * @param previous the previous accessor returned from this method, if available.
+	 * @param onlyExpire true if we are only expiring and not trying to execute anything.
+	 * @return the TxnAccessor of the next scheduled transaction to execute, or null if there are none.
+	 */
+	@Nullable
+	public TxnAccessor triggerNextTransactionExpiringAsNeeded(Instant consensusTime,
+			@Nullable TxnAccessor previous, boolean onlyExpire) {
+
+		LongHashSet seen = null;
+
+		for (long i = 0; i < maxProcessingLoopIterations; ++i) {
+
+			expire(consensusTime);
+
+			var next = store.nextScheduleToEvaluate(consensusTime);
+
+			if (next == null) {
+				return null;
+			}
+
+			// avoid creating a hash set in the normal case where we don't actually process any scheduled txns
+			if (seen == null) {
```

Comment:
What's the reasoning for this test?

The code below seems to guarantee `store.expire(next.getScheduleNum())` will always be called.

Reply:
If you mean the has set creation, then I just wanted to avoid creating the hash set unless it was really going to be used. The vast majority of the time we never get to this line because next is null.

If you mean the set checks in general, then it's just another defensive check to prevent database errors / my errors causing infinite loops or bad state.